Source/msg.cpp

```diff
@@ -1218,44 +1218,34 @@ DWORD OnTargetSpellTile(const TCmd *pCmd, Player &player)
 	return sizeof(message);
 }
 
-DWORD OnOperateObjectTile(const TCmd *pCmd, Player &player)
+size_t OnObjectTileAction(const TCmdLoc &message, Player &player, action_id action, bool pathToTile = true)
 {
-	const auto &message = *reinterpret_cast<const TCmdLocParam1 *>(pCmd);
-	const Point position { message.x, message.y };
+	const Object *object = ObjectAtPosition({ message.x, message.y });
 
-	if (gbBufferMsgs != 1 && player.isOnActiveLevel() && InDungeonBounds(position) && message.wParam1 < MAXOBJECTS) {
-		MakePlrPath(player, position, !Objects[message.wParam1]._oSolidFlag && !Objects[message.wParam1]._oDoorFlag);
-		player.destAction = ACTION_OPERATE;
-		player.destParam1 = message.wParam1;
+	if (gbBufferMsgs != 1 && player.isOnActiveLevel() && object != nullptr) {
+		if (pathToTile)
+			MakePlrPath(player, object->position, !object->_oSolidFlag && !object->_oDoorFlag);
+
+		player.destAction = action;
+		player.destParam1 = object->GetId();
 	}
 
 	return sizeof(message);
 }
 
-DWORD OnDisarm(const TCmd *pCmd, Player &player)
+inline DWORD OnOperateObjectTile(const TCmd *pCmd, Player &player)
 {
-	const auto &message = *reinterpret_cast<const TCmdLocParam1 *>(pCmd);
-	const Point position { message.x, message.y };
-
-	if (gbBufferMsgs != 1 && player.isOnActiveLevel() && InDungeonBounds(position) && message.wParam1 < MAXOBJECTS) {
-		MakePlrPath(player, position, !Objects[message.wParam1]._oSolidFlag && !Objects[message.wParam1]._oDoorFlag);
-		player.destAction = ACTION_DISARM;
-		player.destParam1 = message.wParam1;
-	}
+	return OnObjectTileAction(*reinterpret_cast<const TCmdLoc *>(pCmd), player, ACTION_OPERATE);
+}

```

Comment:
I think it would actually be simple to have `OnObjectTileAction` do the case and then manually inline the 3 calls (completely getting rid of the old functions).